lib/mix/lib/mix/tasks/deps.loadpaths.ex

```diff
@@ -93,25 +93,38 @@ defmodule Mix.Tasks.Deps.Loadpaths do
     end
   end
 
+  defp deps_check(config, all, no_compile?) do
+    with {:compile, _to_compile} <- deps_check(all, no_compile?) do
+      # We need to compile, we first grab the lock, then, we check
+      # again and compile if still applicable
+      Mix.Project.with_build_lock(config, fn ->
```

Comment:
Once the project is compiled, most calls to `deps.loadpaths` actually don't require compilation, so we want want to avoid acquiring a lock upfront.